dbms/src/Flash/Mpp/AsyncRequestHandler.h

```diff
@@ -0,0 +1,389 @@
+            // , so that we can try to connect again.
+            alarm.Set(cq, SteadyClock::now() + std::chrono::seconds(retry_interval_time), this);
+            return true;
+        }
+        else
+        {
+            closeConnection(std::move(done_msg));
+            return false;
+        }
+    }
+
+    GRPCReceiveQueueRes sendPackets()
+    {
+        // note: no exception should be thrown rudely, since it's called by a GRPC poller.
+        while (received_packet_index < read_packet_index)
+        {
+            auto & packet = packets[received_packet_index++];
+            auto res = channel_try_writer.tryWrite<enable_fine_grained_shuffle>(request.source_index, packet);
+            if (res == GRPCReceiveQueueRes::FULL)
+            {
+                packet = std::make_shared<TrackedMppDataPacket>(MPPDataPacketV0);
+                return res;
+            }
+            if (res != GRPCReceiveQueueRes::OK)
+                return res;
+
+            // can't reuse packet since it is sent to readers.
+            packet = std::make_shared<TrackedMppDataPacket>(MPPDataPacketV0);
+        }
+        return GRPCReceiveQueueRes::OK;
+    }
+
+    GRPCReceiveQueueRes reSendPackets()
+    {
+        GRPCReceiveQueueRes res = channel_try_writer.tryReWrite<enable_fine_grained_shuffle>();
+        if (res != GRPCReceiveQueueRes::OK)
+            return res;
+        return sendPackets();
+    }
+
+    bool sendPacketHasError(GRPCReceiveQueueRes res)
+    {
+        return (res == GRPCReceiveQueueRes::CANCELLED || res == GRPCReceiveQueueRes::FINISHED);
+    }
+
+    bool isChannelFull(GRPCReceiveQueueRes res)
+    {
+        return res == GRPCReceiveQueueRes::FULL;
+    }
+
+    // in case of potential multiple inheritances.
+    UnaryCallback<bool> * thisAsUnaryCallback()
+    {
+        return static_cast<UnaryCallback<bool> *>(this);
+    }
+
+    // Do not do anything after this function
+    void asyncWaitForRewrite()
+    {
+        // Do not change the order of these two clauses.
+        // We must ensure that status is set before pushing async handler into queue
+        stage = AsyncRequestStage::WAIT_REWRITE;
+        async_wait_rewrite_queue->push(std::make_pair(&kick_recv_tag, reader->getClientContext()->c_call()));
+    }
+
+    // won't be null and do not delete this pointer
+    grpc::CompletionQueue * cq;
+
+    std::shared_ptr<RPCContext> rpc_context;
+    grpc::Alarm alarm{};
+    Request request;
+
+    String req_info;
+    bool has_data;
+    int retry_times;
+    AsyncRequestStage stage;
+
+    Int32 read_packet_index;
+    Int32 received_packet_index;
+    TrackedMppDataPacketPtrs packets;
+
+    std::shared_ptr<AsyncReader> reader;
+    TrackedMppDataPacketPtr packet;
+    Status finish_status;
+    LoggerPtr log;
+
+    ReceiverChannelTryWriter channel_try_writer;
+    AsyncRequestHandlerWaitQueuePtr async_wait_rewrite_queue;
+    KickReceiveTag kick_recv_tag;
+
+    std::mutex mu;
```

Comment:
Better to set a meaningful name, such as `make_reader_mu`?

Reply:
> Better to set a meaningful name, such as `make_reader_mu`?

okk